### PR TITLE
Added search form and filtering to dumpbag-db lists

### PR DESCRIPTION
**Pull request by smussie**
Adding search functionality for the odoo dump bug database list.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+repos:
```
Comment (p-tombez) [inline .pre-commit-config.yaml:1]: Most of the checks here are related to Odoo addons.
Can you remove or fix their configuration if you want to keep them ?

**Comment by guewen** [comment]
Hello, it's maybe not what was discussed, but I'm still pretty convinced a javascript search is much easier, with better ergonomics (filter at each character typed) and can work on both the list of dumps and the list of databases.

I made a very quick test with this snippet (to copy-paste in the browser's console):

```javascript

function search (text) {
    var items = document.getElementsByTagName('li');
    for (var i = 0; i < items.length; i++) {
        var name = items[i].getElementsByTagName('span')[0].innerHTML;
        if (name.toUpperCase().indexOf(text.toUpperCase()) >= 0) 
            items[i].style.display = 'list-item';
        else
            items[i].style.display = 'none';
    }
}

```

And we can already hide the dumps which does not match a text (`search('blue')`, ...).

Screencast of the JS snippet


What would be missing is:

* Linking the function with an input

```javascript
            <input id="search" type="text" placeholder="Search">

var input = document.getElementById('search');
input.onkeyup = function() {
    search(input.value);
}

```

* hiding the `<ul>` when no `<li>` is visible
* ensuring the function is active only on the `<li>` under a `<ul>` with class `dump-list` (to ensure we don't hide menus or whatever)
* optionally, adding the same thing for databases

**Comment by guewen** [comment]
Quick and dirty screencast using the js from my comment above in the browser inspector:

Screencast of filtering

File: .pre-commit-config.yaml (hunk 1 above)
Comment (guewen) [reply .pre-commit-config.yaml:1]: In fact there is 2 pull requests for this change, you can update an already existing one.
I proposed a configuration to use here: https://github.com/camptocamp/odoo-dump-bag/pull/16#pullrequestreview-430419816

File: server/dumpbagserver/views.py [hunk 2]
```diff
@@ -5,55 +5,73 @@
...
 from werkzeug.routing import UnicodeConverter, ValidationError
 
 from .bagger import Bagger
+from .forms import SearchForm
 
 
 class DBNameConverter(UnicodeConverter):
-
     def to_python(self, value):
         if not re.match("^[A-Za-z0-9_-]+$", value):
```
[the remark this answers is not in the document]
Comment (chroi) [reply server/dumpbagserver/views.py:17]: Can it be the precommit scripts that changed the quotes ?

[the remark this answers is not in the document]
Comment (p-tombez) [reply server/dumpbagserver/views.py:17]: Yes, it's `black`

File: server/dumpbagserver/views.py [hunk 3]
```diff
@@ -5,55 +5,73 @@
...
+            download_commands=bagger.download_commands,
+            form=search_form,
+        )
+    else:
+        return render_template(
+            "dumps.html",
+            dumps=dumps,
+            download_commands=bagger.download_commands,
+            form=search_form,
+        )
```
Comment (guewen) [inline server/dumpbagserver/views.py:57]: the duplication of the call to `render_template` is not necessary:

```python
if request.args.get("search", type=str) and search_form.validate():
    sname = request.args.get("search", type=str)
    dumps = {dbname: value for dbname, value in dumps.copy().items() if sname in dbname}
return render_template(
    "dumps.html",
    dumps=dumps,
    download_commands=bagger.download_commands,
    form=search_form,
)
```

File: server/dumpbagserver/forms.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    def validate_search(form, field):
+        bagger = Bagger(app_config)
+        dumps = bagger.list_dumps()
+        parse = form.search.data
+        find = False
+        for bds in dumps:
+            if parse in bds:
+                find = True
+        if find is False:
+            raise ValidationError(parse + " not in any of the db names.")
```
Comment (guewen) [inline server/dumpbagserver/forms.py:27]: Is it necessary to raise an error rather than returning an empty list? There is code duplication here with what's done in the `/` route and it seems unnecessary at a first glance.
And even if we want an error, I think a validator is not the right place to do it, the validator validates the "input", not the output, if we want an error, it should be raised by the endpoint `/`.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (smussie) [reply .pre-commit-config.yaml]: @p-tombez You are right, thanks. thus we removed the pre-commit-config.yaml.

File: server/dumpbagserver/views.py (hunk 3 above)
Comment (smussie) [reply server/dumpbagserver/views.py]: @guewen Is is done, thanks.

Comment (smussie) [reply server/dumpbagserver/views.py]: @guewen We added also your suggestion with the javascript. It is very much appreciated, thank you.

File: server/dumpbagserver/views.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (smussie) [reply server/dumpbagserver/views.py:17]: @mmequignon Sorry, all the changes were related to pre-commit reformat before.Thanks.

File: server/dumpbagserver/views.py (hunk 3 above)
Comment (guewen) [reply server/dumpbagserver/views.py]: hmm but what's the reason to have both the server-side search *and* the js search? My remark was that using a server-side search is overkill for the need and less ergonomic

Comment (guewen) [reply server/dumpbagserver/views.py]: I point out that my js code was incomplete (https://github.com/camptocamp/odoo-dump-bag/pull/17#issuecomment-650166475 for what's missing), mainly, it should hide the main titles of databases when no database is found. But if you need help on that, let me know.

Comment (chroi) [reply server/dumpbagserver/views.py]: Having both makes the service degrade gracefully (and doesn't hurt otherwise).

File: server/dumpbagserver/forms.py (hunk 4 above)
Comment (guewen) [reply server/dumpbagserver/forms.py:27]: This `validate_search` method is to remove: if you want to raise an error, do it in the controller's endpoint.
You should not do the logic of the endpoint in the validor, only validates if the user input is valid.

File: server/dumpbagserver/forms.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+from .bagger import Bagger
+
+
+class SearchForm(Form):
+    search = StringField(
+        "Search",
+        validators=[
+            DataRequired(),
+            Regexp(r"^\w+$", message="Only letters numbers or _."),
```
Comment (guewen) [inline server/dumpbagserver/forms.py:13]: We should add `-`

File: server/dumpbagserver/views.py (hunk 3 above)
Comment (guewen) [reply server/dumpbagserver/views.py]: I would say that every line of code hurts a bit (code is a liability), but whatever, if you want.
Still, note that as of today, if you disable javascript, you can't "unroll" the sections, so you won't go far. A fix could be to show everything by default when js is disabled, but then CTRL-F would be a good enough graceful search :stuck_out_tongue_closed_eyes:

File: server/dumpbagserver/templates/dumps.html [hunk 6]
```diff
@@ -63,9 +60,22 @@ <h2 class="dump-title" onclick="toggleGroup('dump-list-for-{{database}}', '{{dat
   if (hash) {
     toggleVisibility('dump-list-for-' + hash);
   }
+  function search (text) {
+      var items = document.getElementsByTagName('li');
+      for (var i = 0; i < items.length; i++) {
```
Comment (guewen) [inline server/dumpbagserver/templates/dumps.html:65]: in order to hide the database names when no database is found:

* hide all the `<h2 class="dump-title"/>` before the for loop
* inside the for loop, when we find a dump matching the search, show the parent `<h2 class="dump-title"/>`

```javascript
        if (name.toUpperCase().indexOf(text.toUpperCase()) >= 0) {
              items[i].style.display = 'list-item';
              // here, find the parent h2.dump-title and show it
         }
```

File: server/dumpbagserver/templates/dumps.html [hunk 7]
```diff
@@ -63,9 +60,22 @@ <h2 class="dump-title" onclick="toggleGroup('dump-list-for-{{database}}', '{{dat
   if (hash) {
     toggleVisibility('dump-list-for-' + hash);
   }
+  function search (text) {
+      var items = document.getElementsByTagName('li');
```
Comment (guewen) [inline server/dumpbagserver/templates/dumps.html:64]: restrict to elements inside `document.querySelectorAll('.dump-list')`

File: .pre-commit-config.yaml (hunk 1 above)
Comment (guewen) [reply .pre-commit-config.yaml]: The configuration I proposed on https://github.com/camptocamp/odoo-dump-bag/pull/16#pullrequestreview-430419816
was fine though.

**Comment by smussie** [comment]
dumpbag

File: server/dumpbagserver/templates/dumps.html [hunk 8]
```diff
@@ -1,44 +1,43 @@
 {% extends "layout.html" %}
 {% set active_page = "dumps" %}
 {% block body %}
-
-<div class="dumps">
```
Comment (guewen) [inline server/dumpbagserver/templates/dumps.html:5]: is this removal on purpose? If yes, there is an unmatched `</div>` at the end of the file

Comment (smussie) [reply server/dumpbagserver/templates/dumps.html:5]: Screenshot from 2020-07-13 16-20-12

Comment (smussie) [reply server/dumpbagserver/templates/dumps.html:5]: This was in the style so it was an oversight on my part, i added it back. Thank you.
